text/3735-field-projections.md

```diff
@@ -0,0 +1,1344 @@
+[`field-projection`]: https://crates.io/crates/field-projection
+
+## Blog Posts and Discussions
+
+- [Safe Cell field projection in
+  Rust](https://www.abubalay.com/blog/2020/01/05/cell-field-projection)
+- [Field projdection for `Rc` and
+  `Arc`](https://internals.rust-lang.org/t/field-projection-for-rc-and-arc/15827)
+- [Generic Field Projection](https://internals.rust-lang.org/t/generic-field-projection/16204)
+
+Blog posts about pin (projections):
+- [Pinned places](https://without.boats/blog/pinned-places/)
+- [Overwrite trait](https://smallcultfollowing.com/babysteps/series/overwrite-trait/)
+
+## Rust and Other Languages
+
+Rust already has a precedent for compiler-generated types. All functions and closures have a unique,
+unnameable type.
+
+In C++ there are field projections supported on `std::shared_ptr`, it consists of two pointers, one
+pointing to the reference count and the other to the data. Making it possible to project down to a
+field and still take a reference count on the entire struct, keeping also the field alive.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+## Syntax Bikeshedding
+
+What is the right syntax for the various operations given in this RFC?
+
+Ideally we would have a strong opinion when this feature is implemented. But the decision should
+only be finalized when stabilizing the feature.
+
+### Field Projection Operator
+
+Current favorite: `$base:expr->$field:ident`.
+
+Alternatives:
+- use `~` instead of `->`
+
+### Naming Field Types
+
+Current favorite: `field_of!($Container:ty, $($fields:expr)+ $(,)?)` macro with `offset_of!` syntax.
+
+Alternatives:
+- Introduce a more native syntax on the level of types `$Container:ty->$field:ident` akin to
+  projecting an expression.
+
+### Declaring a Transparent Container Type
+
+Current favorite: [`#[projecting]`](#projecting-attribute) attribute.
+
+Alternatives:
+- extend the `repr` attribute with
+  - `projecting`
+  - `flatten`
+- use `#[flatten]` instead.
+
+## Other
+
+- should the [`#[projecting]`](#projecting-attribute) attribute have an associated field attribute
+  to mark the field that is projected onto?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Enums
+
+Enums are difficult to support with the same framework as structs. The problem is that many
+containers don't provide sufficient guarantees to read the discriminant (for example raw pointers
+and `&mut MaybeUninit<T>`). However, for types that do provide sufficient guarantees, one could cook
+up a similar feature. Let's call them *enum projections*. They could work like this: projecting is
+done via a new kind of match operator:
+
+```rust
+enum MyEnum<F> {
+    A(i32, String),
+    B(#[pin] F),
+}
+type F = impl Future;
+let x: Pin<&mut MyEnum<F>>;
+match_proj x {
+    MyEnum::A(n, s) => {
+        let _: &mut i32 = n;
+        let _: &mut String = s;
+    }
+    MyEnum::B(fut) => {
+        let _: Pin<&mut F> = fut;
+    }
+}
```

Comment:
This is very similar to what match ergonomics do today: an outer `&` or `&mut` "slips into" the fields when matching. I think language-wise it would work well to extend the normal `match` statement to work with projectable pointers, maybe with a special `unsafe trait EnumProjectable` that allows retrieving the discriminant and guarantees we can project the fields simultaneously.

Reply:
What about the case, where the container itself is an enum? One probably wants to be able to both match the raw enum wrapper **and** the projected enum inside:
```rust
let x: Cow<'_, MyEnum<F>>;
match x {
    Borrowed(x) => { /* ... */ }
    Owned(x) => { /* ... */ }
}
match_proj x {
    MyEnum::A(n, s) => {
        let _: Cow<'_, i32> = n;
        let _: Cow<'_, String> = s;
    }
    MyEnum::B(fut) => {
        let _: Cow<'_, F> = fut;
    }
}
```

Reply:
We've been having this exact issue in the design of [deref patterns](https://hackmd.io/4qDDMcvyQ-GDB089IPcHGg?view=) (which aims to allow this code but give you `n: &i32` and `s: &String`). This particular case isn't ambiguous because we know that `MyEnum::A` is of type `MyEnum`, then type checking will see the mismatch between `MyEnum` and `Cow` and attempt to project. So we can accept both.

There is actual ambiguity if you have a `Cow<Cow<T>>`, but the little algorithm I just sketched would simply resolve to the outer `Cow`.

Reply:
> We've been having this exact issue in the design of [deref patterns](https://hackmd.io/4qDDMcvyQ-GDB089IPcHGg?view=) (which aims to allow this code but give you `n: &i32` and `s: &String`). This particular case isn't ambiguous because we know that `MyEnum::A` is of type `MyEnum`, then type checking will see the mismatch between `MyEnum` and `Cow` and attempt to project. So we can accept both.

That's pretty clever!

> There is actual ambiguity if you have a `Cow<Cow<T>>`, but the little algorithm I just sketched would simply resolve to the outer `Cow`.

I see, for the particular `Cow` case, I wouldn't expect people to use nested cows... And other pointer-like enum wrappers should also only rarely be nested, so seems like a reasonable compromise.

I really like how this makes enum projections this simple, I am thinking about adding them to the main text of the RFC. Since without enum projections, one cannot easily design a `MaybeDone` future. And it would be really nice to have pin projections also for enums.

I will see if someone complains and if not, I will move it into the main section.

Reply:
Given how deref patterns is stuck in design limbo, I would propose you leave this to the "Future work" section. I'll definitely incorporate discussion on this into the deref patterns discussions tho.